src/Orleans.CodeGeneration/ApplicationPartManagerCodeGenExtensions.cs

```diff
@@ -0,0 +1,45 @@
+﻿using Microsoft.Extensions.Logging.Abstractions;
+using Orleans.ApplicationParts;
+using Orleans.CodeGenerator;
+using Orleans.Metadata;
+using Orleans.Runtime;
+using Orleans.Serialization;
+
+namespace Orleans.Hosting
+{
+    /// <summary>
+    /// Extensions for <see cref="IApplicationPartManagerWithAssemblies"/> for invoking code generation.
+    /// </summary>
+    public static class ApplicationPartManagerCodeGenExtensions
+    {
+        /// <summary>
+        /// Generates support code for the the provided assembly and adds it to the builder.
+        /// </summary>
+        /// <param name="manager">The builder.</param>
+        /// <returns>A builder with support parts added.</returns>
+        public static IApplicationPartManagerWithAssemblies WithCodeGeneration(this IApplicationPartManagerWithAssemblies manager)
+        {
+            var tempPartManager = new ApplicationPartManager();
+            foreach (var provider in manager.FeatureProviders)
+            {
+                tempPartManager.AddFeatureProvider(provider);
+            }
+
+            foreach (var part in manager.ApplicationParts)
+            {
+                tempPartManager.AddApplicationPart(part);
+            }
+
+            tempPartManager.AddApplicationPart(new AssemblyPart(typeof(RuntimeVersion).Assembly));
+            tempPartManager.AddFeatureProvider(new AssemblyAttributeFeatureProvider<GrainInterfaceFeature>());
+            tempPartManager.AddFeatureProvider(new AssemblyAttributeFeatureProvider<GrainClassFeature>());
+            tempPartManager.AddFeatureProvider(new AssemblyAttributeFeatureProvider<SerializerFeature>());
+            tempPartManager.AddFeatureProvider(new BuiltInTypesSerializationFeaturePopulator());
+            
+            var codeGenerator = new RoslynCodeGenerator(tempPartManager, new NullLoggerFactory());
+            var generatedAssembly = codeGenerator.GenerateAndLoadForAssemblies(manager.Assemblies);
```

Comment:
Does it mean that codegen is run immediately? Shouldn't it wait until a specific point in time just before building the container, so that all the different codegen tasks occur together? Because if I use the WithCodeGeneration extension method multiple times, it will run multiple times. Or did I misunderstand this?

Reply:
That's correct - code gen will run immediately, but not necessarily for one assembly at a time. The user has a degree of control over it.

In the following code, codegen is executed twice: once for IMyGrain's asm and MyGrain's asm and then again for whatever asms are in the app base dir. The second execution will see the code generated by the first, so it won't double-generate.
```C#
.ConfigureApplicationPartManager(parts =>
{
    parts.AddApplicationPart(typeof(IMyGrain).Assembly)
         .AddApplicationPart(typeof(MyGrain).Assembly)
         .WithCodeGeneration();
    parts.AddFromApplicationBaseDirectory().WithCodeGeneration();
});
```

Does that seem alright or would you prefer that all actions are queued up and executed only during Build()?

Reply:
I guess it's fine if there is a way to chain everything and then execute once. And if they don't double-generate even when executing twice, then that's harmless anyway.